Modules/FireForget/Get-AbortCleanKansaServersFF.ps1

```diff
@@ -0,0 +1,184 @@
+# This module is used to abort the deployment of a Kansa module from distributed-kansa servers
+# It can also be used in the event that a kansa server stalls its deployment due to too many connection
+# errors or timeouts.  The parameters can be configured to clean up result folders and error logs
+# and optionally send any uncollected results/errors to ELK for the previous n kansa jobs.
+# The default settings are set for safe operation and will only send results/errors from the last 1 job
+
+if(!(Get-Variable -Name KillRunningKansaJob -ErrorAction SilentlyContinue)){$KillRunningKansaJob = $False}
+if(!(Get-Variable -Name CleanupServer -ErrorAction SilentlyContinue)){$CleanupServer = $False}
+if(!(Get-Variable -Name SendPartialResultsToElk -ErrorAction SilentlyContinue)){$SendPartialResultsToElk = $True}
+if(!(Get-Variable -Name DeletePartialResults -ErrorAction SilentlyContinue)){$DeletePartialResults = $False}
+if(!(Get-Variable -Name NumRecentJobs -ErrorAction SilentlyContinue)){$NumRecentJobs=1}
+if(!(Get-Variable -Name KansaRemotePath -ErrorAction SilentlyContinue)){$KansaRemotePath="F:\Kansa"}
+if(!(Get-Variable -Name KansaJobPath -ErrorAction SilentlyContinue)){$KansaJobPath = ""}
+
+$thisProcess = Get-WmiObject -query 'Select * from win32_process' | where ProcessId -eq $Pid | Select Name,ProcessId,ParentProcessId
+
+$allPaths = @()
+if($KansaJobPath -eq ""){
+    $KansaJobPath = $KansaRemotePath
+    if ($NumRecentJobs -eq 0) {
+        $allPaths = gci -Path $KansaJobPath -Directory -Recurse | where Name -Match "Output_" | Sort-Object -Descending
+    } 
+    else {
+        $allPaths = gci -Path $KansaJobPath -Directory -Recurse | where Name -Match "Output_" | Sort-Object -Descending | Select -First $NumRecentJobs
+    }
+}
+else {
+    $allPaths = $KansaJobPath
+}
+
+if ($KillRunningKansaJob) {
+    $thePIDSkilled = New-Object -TypeName System.Collections.ArrayList
+    gps | where ProcessName -eq "powershell" | ForEach-object {
```

Comment:
Does this kill any powershell process? Even if it's not part of a Kansa job?

Reply:
Yes,  This module is intended to be launched against a "Kansa-Server" which is one of the designated servers used solely for the launching of distributed kansa jobs at scale.  In the event that something goes wrong, or network/enterprise availability necessitates the cancellation of a deployment in-progress we built this module to terminate active Fire&Forget deployments on distributed Kansa servers.  The module also has the ability to reset the underlying services by bouncing WinRM, Windows Management (WMI), etc.. and send any partial results into logging collectors (i,e ELK).  

In our Fire&Forget wrapper we do have a killswitch that terminates powershell processes filtering on the account/user-context that launched them.  We could add that same restriction here if you'd like.  We just wanted to make sure this "panic-button" to abort a kansa-deployment against 200K endpoints would have the minimum number of filters that might prevent it from killing an active powershell process on a distributed kansa server

Reply:
Ah, so this isn't aimed at individual systems, but at servers running Kansa. Makes sense.